### PR TITLE
feat: WebUI group list page [DET-7921, DET-7976]

### PR DESCRIPTION
**Pull request by gt2345**
## Description

Implement group list page in Web UI. 

## Test Plan

Navigate to `/det/settings/group-management/rbac-enabled`, create/edit/delete several groups.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        label={GROUP_NAME_LABEL}
+        name={GROUP_NAME_NAME}
+        required
+        rules={[
+          {
+            message: 'Please type in your group name.',
+            required: true,
+          },
+        ]}
+        validateTrigger={[ 'onSubmit' ]}>
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx:67]: also use `onChange`?

File: webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      // Re-throw error to prevent modal from getting dismissed.
+      throw e;
+    }
+  }, [ onClose, group ]);
+
+  const modalOpen = useCallback(() => {
+    openOrUpdate({
+      closable: true,
+      content:
+      `Are you sure you want to delete group ${group.group.name} (ID: ${group.group.groupId}).`,
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx:40]: `?` not `.`

File: webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  const [ form ] = Form.useForm();
+
+  const { modalOpen: openOrUpdate, ...modalHook } = useModal();
+
+  const handleCancel = useCallback(() => {
+    form.resetFields();
+  }, [ form ]);
+
+  const handleOkay = useCallback(async () => {
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx:120]: supernit: `handleOk` to match `onOk`

File: webui/react/src/pages/Settings/GroupManagement.tsx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+      message.error('error deleting group');
+      handleError(e, { silent: true, type: ErrorType.Input });
+    }
+  }, [ onExpand, fetchGroups ]);
+
+  const expandedUserRender = useCallback((record:V1GroupSearchResult) => {
+    const { group: { groupId } } = record;
+    const g = groups.find((gr) => gr.group.group.groupId === groupId);
+    const userColumn = [ {
+      dataIndex: 'displayName',
```
Comment (johnkim-det) [inline webui/react/src/pages/Settings/GroupManagement.tsx:175]: bug: not displaying `displayName` column

File: webui/react/src/pages/Settings/GroupManagement.tsx [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+      message.success('User has been deleted.');
+      onExpand(true, record);
+      fetchGroups();
+    } catch (e) {
+      message.error('error deleting group');
+      handleError(e, { silent: true, type: ErrorType.Input });
+    }
+  }, [ onExpand, fetchGroups ]);
+
+  const expandedUserRender = useCallback((record:V1GroupSearchResult) => {
```
Comment (johnkim-det) [inline webui/react/src/pages/Settings/GroupManagement.tsx:171]: bug: when expanded and group is edited, the expanded table gets cleared and shows "No Data"

File: webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+      throw e;
+    }
+  }, [ form, onClose, group ]);
+
+  const modalOpen = useCallback(() => {
+    openOrUpdate({
+      closable: true,
+      content: <ModalForm form={form} group={group} users={users} />,
+      icon: null,
+      okText: group ? MODAL_HEADER_LABEL_EDIT : MODAL_HEADER_LABEL_CREATE,
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx:148]: nit: i'd prefer something like "Save Group" and/or "Save Changes" for the button, but reusing the modal title is fine too, really.

File: webui/react/src/pages/Settings/GroupManagement.tsx [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+      width: '40%',
+    }, {
+      dataIndex: 'username',
+      key: 'username',
+      title: 'User Name',
+      width: '50%',
+    }, {
+      key: 'action',
+      render: (_:string, r: V1User) => (
+        <Button onClick={() => onRemoveUser(record, r.id)}>Remove</Button>),
```
Comment (johnkim-det) [inline webui/react/src/pages/Settings/GroupManagement.tsx:187]: bug: removing a record also clears the entire expanded table

File: webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+interface ModalProps {
+  group: V1GroupSearchResult;
+  onClose?: () => void;
+}
+
+const useModalDeleteGroup = ({ onClose, group }: ModalProps): ModalHooks => {
+
+  const { modalOpen: openOrUpdate, ...modalHook } = useModal();
+  const handleOkay = useCallback(async () => {
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx:21]: super nit: ditto `handleOk`

File: webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.test.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+  return view;
+};
+
+describe('useModalCreateGroup', () => {
+  it('should open modal with correct values', async () => {
+    await setup();
+
+    expect(screen.getByText(
+      'Are you sure you want to delete group test_groupname1 (ID: 1).',
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.test.tsx:66]: use exported variable?

File: webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    if (!group.group.groupId) return;
+    try {
+      await deleteGroup({ groupId: group.group.groupId });
+      message.success(API_SUCCESS_MESSAGE);
+      onClose?.();
+    } catch (e) {
+      message.error('error deleting group');
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx:28]: nit: sentence case?

File: webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        await updateGroup({ groupId: group.group.groupId, ...formData });
+        message.success(API_SUCCESS_MESSAGE_EDIT);
+      } else {
+        await createGroup(formData);
+        message.success(API_SUCCESS_MESSAGE_CREATE);
+      }
+      form.resetFields();
+      onClose?.();
+    } catch (e) {
+      message.error('error creating new group');
```
Comment (johnkim-det) [inline webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx:135]: nit: sentence case?

**Review by johnkim-det — COMMENTED** [review]
saw some bugs. might be specific to my local user data, i'm not sure. lmk if you need help reproducing.

File: webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx (hunk 1 above)
Comment (gt2345) [reply webui/react/src/hooks/useModal/UserSettings/useModalCreateGroup.tsx:67]: Makes sense

File: webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx (hunk 2 above)
Comment (gt2345) [reply webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.tsx:40]: fixed

File: webui/react/src/pages/Settings/GroupManagement.tsx (hunk 5 above)
Comment (gt2345) [reply webui/react/src/pages/Settings/GroupManagement.tsx:171]: This has been fixed

File: webui/react/src/pages/Settings/GroupManagement.tsx (hunk 7 above)
Comment (gt2345) [reply webui/react/src/pages/Settings/GroupManagement.tsx:187]: Can't believe I missed this... fixed~

File: webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.test.tsx (hunk 9 above)
Comment (gt2345) [reply webui/react/src/hooks/useModal/UserSettings/useModalDeleteGroup.test.tsx:66]: fixed

**Review by johnkim-det — COMMENTED** [review]
fixes look good, one last thing though is that I think this should use the `getDisplayName` util function instead of accessing the users' `displayName` field directly, in case it doesn't exist and we want to use `username` instead.

**Review by johnkim-det — APPROVED** [review]
> > fixes look good, one last thing though is that I think this should use the `getDisplayName` util function instead of accessing the users' `displayName` field directly, in case it doesn't exist and we want to use `username` instead.
> 
> @johnkim-det Thank you! Just to make sure, you meant using `getDisplayName` in drop down, right? Since we have separate columns for `Display Name` and `User Name`, if display name does not exist, I think that column should be left empty

was thinking both but i didn't notice that there was a separate column for usernames, so that makes sense

**Pull request merged**